meinberlin/apps/projects/tasks.py

```diff
@@ -0,0 +1,72 @@
+import redis
+from celery import shared_task
+from django.utils import timezone
+
+from adhocracy4.phases.models import Phase
+from meinberlin.apps import logger
+
+redis_db = redis.Redis()
+
+
+@shared_task  # periodic task every 10min
+def schedule_reset_cache_for_future_projects():
+    # fetch earliest future project
+    # check if redis has a next_project_starts key
+    next_project_starts = redis_db.get("next_project_starts")
+    
+    if not next_project_starts:
+        # query the db 
+        # retreive the next_project_starts from db'
+        phase = Phase.objects.filter(start_date__isnull=False).first()
+        if not phase:
+            logger.warning("At least one phase with dates is required")
+            return
+
+        next_project_starts = phase.start_date.strftime("%Y-%m-%d, %H:%M:%S")
+        # set the redis key: value
+        redis_db.mset({"next_project_starts": next_project_starts})
+
+    # compare now with next start date
+    project_starts = datetime.strptime(next_project_starts, "%Y-%m-%d, %H:%M:%S")
+    remain_time = project_starts - timezone.now()
+    if  remain_time <= timedelta(minutes=10):
+        # schedule cache clear for the remaining time btwn now and next start
+        reset_cache_for_projects.apply_async((starts=True), countdown=remain_time)
+    else: 
+        logger.info("No cache reset for future projects")
+
+
+@shared_task  # periodic task every 10min
+def schedule_reset_cache_for_past_projects():
+    # fetch latest past project
+    # check if redis has a next_project_ends key
+    next_project_ends = redis_db.get("next_project_ends")
+    if not next_project_ends:
+        # query the db 
+        # retreive the next_project_ends from db
+        phase = Phase.objects.filter(start_date__isnull=False).first()
+        if not phase:
+            logger.warning("At least one phase with dates is required")
+            return
+        next_project_ends = phase.end_date.strftime("%Y-%m-%d, %H:%M:%S")
+        # set the redis key: value
+        redis_db.mset({"next_project_ends": next_project_ends})
+
+    # compare now with next end date
+    project_ends = datetime.strptime(next_project_ends, "%Y-%m-%d, %H:%M:%S")
+    remain_time = timezone.now() - project_ends
+    if  remain_time <= timedelta(minutes=10):
+        # schedule cache clear for the remaining time btwn now and next end
+        reset_cache_for_projects.apply_async((ends=True), countdown=remain_time)
+
+
+@shared_task
+def reset_cache_for_projects(ends=False, starts=False):
+    # cache set projects
+    if starts:
+        # remove redis key next_project_start
+        redis_db.delete("next_project_starts")
+    if ends:
+        # remove redis key next_project_ends
+        redis_db.delete("next_project_ends")
+
```

Comment:
@goapunk have a look and we can discuss it again TMR. The logic is to check every 10 min two things:
one task for future and one for past projects:  if the earliest future/latest past project is happening in the next 10 mins, and if yes then we call the actual reset cache task with a countdown. If not, then one of the next callings of the periodic tasks will do it.

We only query the DB when there is no redis key set, and when a project is saved.

Reply:
structure looks good to me, left a small question but I think this will be a nice solution